File: python/algorithms/sorting/quick_sort/randomized/quick_sort_randomized.py
import random 

def quick_sort(arr, low, high):
    if low < high:
        p = randomized_partition(arr, low, high)

        quick_sort(arr, low, p - 1)     # left side
        quick_sort(arr, p + 1, high)    # right side
    
def randomized_partition(arr, low, high):
    pivot_index = random.randint(low, high)
    arr[pivot_index], arr[high] = arr[high], arr[pivot_index]
    return partition(arr, low, high)

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

# Sample of use
if __name__ == "__main__":
    A = [2, 8, 7, 1, 3, 5, 6]
    print("Before:", A)
    quick_sort(A, 0, len(A) - 1)
    print("After:", A)

File: java/algorithms/sorting/quick_sort/randomized/QuickSortRandomized.java
package algorithms.sorting.quick_sort.randomized;
import java.util.Random;


public class QuickSortRandomized {

    public static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int p = randomizedPartition(arr, low, high);

            // recursive calls
            quickSort(arr, low, p - 1);   // left side
            quickSort(arr, p + 1, high);  // right side
        }
    }

    private static int randomizedPartition(int[] arr, int low, int high) {
        Random rand = new Random();
        int pivotIndex = rand.nextInt(high - low + 1) + low; // random between low and high
        swap(arr, pivotIndex, high);
        return partition(arr, low, high);
    }

    private static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;

        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                swap(arr, i, j);
            }
        }

        swap(arr, i + 1, high);
        return i + 1;
    }

    // Swap helper method
    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    // Sample usage
    public static void main(String[] args) {
        int[] A = {2, 8, 7, 1, 3, 5, 6};
        System.out.println("Before: ");
        printArray(A);

        quickSort(A, 0, A.length - 1);

        System.out.println("After: ");
        printArray(A);
    }

    // Helper to print arrays
    private static void printArray(int[] arr) {
        for (int x : arr) {
            System.out.print(x + " ");
        }
        System.out.println();
    }
}